csharp/src/Ice/Reference.cs

```diff
@@ -1300,6 +1300,11 @@ internal void Write(OutputStream ostr)
                 throw new ArgumentException(
                     $"invocation mode `{InvocationMode}' is not compatible with the ice2 protocol");
             }
+
+            if (Endpoints.FirstOrDefault(endpoint => endpoint.Protocol != Protocol) is Endpoint endpoint)
+            {
+                throw new ArgumentException($"endpoint `{endpoint}' is not compatible with {Protocol} protocol");
```

Comment:
This could/should be an assert, not a runtime check. For most calls to this ctor (like from Parse), the just-created endpoints were created with Protocol.

I believe this check breaks ObjectAdapter.CreateReference, which needs to be fixed.

I would move the check to Clone, i.e. when the application provides endpoints to Clone, they must use the correct protocol.

Reply:
what is the issue with `ObjectAdapter.CreateReference`?

I think we need a similar check for `SetPublishedEndpointsAsync`, the published endpoints protocol should match the adapter protocol

Reply:
The published endpoints may not match the OA endpoint. For example, you're publishing the endpoints of a "router" that listens on ice1 but forwards to your ice2 object adapter.

In CreateReference, the proxy's protocol should be the published endpoints' protocol (when published endpoints are set) as opposed to the OA's protocol.

Reply:
Should we check that all published endpoints use the same protocol?

Reply:
The way the OA parses/creates the published endpoints, they have necessarily al the same protocol.

> A related question is if we should throw when set published endpoints in a nameless OA, I think this should be an error.
I think it's fine to throw in this situation, although we may have a test that relies on this odd behavior.

Reply:
I mean when calling `public void SetPublishedEndpoints(IEnumerable<Endpoint> newEndpoints)`